File: internal/handler/logging/request_payload_logger.go
package logging

import (
	"net/http"

	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
)

// RequestPayloadLogger handles logging of request payload information
type RequestPayloadLogger struct {
	cloudflareHelper CloudflareHelper
	correlationID    CorrelationID
	applicationContext ApplicationContext
}

// CloudflareHelper defines the interface for Cloudflare-related operations
type CloudflareHelper interface {
	StoreCloudflareHeaders(payload map[string]interface{}, r *http.Request)
}

// CorrelationID defines the interface for correlation ID operations
type CorrelationID interface {
	CurrentID() string
}

// ApplicationContext defines the interface for application context operations
type ApplicationContext interface {
	Current() map[string]interface{}
}

// NewRequestPayloadLogger creates a new instance of RequestPayloadLogger
func NewRequestPayloadLogger(
	cloudflareHelper CloudflareHelper,
	correlationID CorrelationID,
	applicationContext ApplicationContext,
) *RequestPayloadLogger {
	return &RequestPayloadLogger{
		cloudflareHelper: cloudflareHelper,
		correlationID: correlationID,
		applicationContext: applicationContext,
	}
}

// AppendInfoToPayload appends information to the payload for logging
func (r *RequestPayloadLogger) AppendInfoToPayload(payload map[string]interface{}, req *http.Request, authUser *model.User, urgency *model.RequestUrgency) map[string]interface{} {
	// Add user agent
	payload["ua"] = req.Header.Get("User-Agent")

	// Add remote IP
	payload["remote_ip"] = req.RemoteAddr

	// Add correlation ID
	payload["correlation_id"] = r.correlationID.CurrentID()

	// Add application context metadata
	payload["metadata"] = r.applicationContext.Current()

	// Add request urgency and target duration if defined
	if urgency != nil {
		payload["request_urgency"] = urgency.Name
		payload["target_duration_s"] = urgency.Duration
	}

	// Add user ID and username if a user is logged in
	if authUser != nil {
		payload["user_id"] = authUser.ID
		payload["username"] = authUser.Username
	}

	// Add queue duration
	payload["queue_duration_s"] = req.Header.Get("X-Gitlab-Rails-Queue-Duration")

	// Store Cloudflare headers
	r.cloudflareHelper.StoreCloudflareHeaders(payload, req)

	return payload
}

File: rust_app/src/controllers/concerns/request_payload_logger.rs
use crate::{
    config::Settings,
    models::User,
    utils::{
        application_context::ApplicationContext, cloudflare::CloudflareHelper,
        correlation::CorrelationId,
    },
};
use actix_web::{dev::ServiceRequest, web::Data};
use serde_json::Value;
use std::collections::HashMap;

pub trait RequestPayloadLogger {
    fn append_info_to_payload(&self, payload: &mut HashMap<String, Value>, req: &ServiceRequest);
}

pub struct RequestPayloadLoggerImpl {
    settings: Data<Settings>,
}

impl RequestPayloadLoggerImpl {
    pub fn new(settings: Data<Settings>) -> Self {
        Self { settings }
    }
}

impl RequestPayloadLogger for RequestPayloadLoggerImpl {
    fn append_info_to_payload(&self, payload: &mut HashMap<String, Value>, req: &ServiceRequest) {
        // Add user agent
        if let Some(user_agent) = req.headers().get("user-agent") {
            if let Ok(ua) = user_agent.to_str() {
                payload.insert("ua".to_string(), Value::String(ua.to_string()));
            }
        }

        // Add remote IP
        if let Some(remote_addr) = req.connection_info().peer_addr() {
            payload.insert(
                "remote_ip".to_string(),
                Value::String(remote_addr.to_string()),
            );
        }

        // Add correlation ID
        payload.insert(
            "correlation_id".to_string(),
            Value::String(CorrelationId::current().to_string()),
        );

        // Add application context
        if let Some(ctx) = ApplicationContext::current() {
            payload.insert(
                "metadata".to_string(),
                serde_json::to_value(ctx).unwrap_or(Value::Null),
            );
        }

        // Add user information if available
        if let Some(user) = req.extensions().get::<User>() {
            payload.insert("user_id".to_string(), Value::Number(user.id.into()));
            payload.insert("username".to_string(), Value::String(user.username.clone()));
        }

        // Add queue duration if available
        if let Some(duration) = req.extensions().get::<f64>() {
            payload.insert(
                "queue_duration_s".to_string(),
                Value::Number((*duration).into()),
            );
        }

        // Add Cloudflare headers
        CloudflareHelper::store_headers(payload, req);
    }
}
